boa_engine/src/syntax/ast/node/visit.rs

```diff
@@ -0,0 +1,2317 @@
+use crate::syntax::ast::node::operator::assign::AssignTarget;
+use crate::syntax::ast::node::template::TemplateElement;
+use crate::syntax::ast::node::{
+    ArrayDecl, ArrowFunctionDecl, Assign, AsyncFunctionDecl, AsyncFunctionExpr, AsyncGeneratorDecl,
+    AsyncGeneratorExpr, AwaitExpr, BinOp, Block, Break, Call, Case, Catch, Class, ConditionalOp,
+    Continue, Declaration, DeclarationList, DeclarationPattern, DoWhileLoop, Finally, ForInLoop,
+    ForLoop, ForOfLoop, FormalParameter, FormalParameterList, FormalParameterListFlags,
+    FunctionDecl, FunctionExpr, GeneratorDecl, GeneratorExpr, GetConstField, GetField,
+    GetPrivateField, GetSuperField, Identifier, If, New, Object, Return, Spread, StatementList,
+    SuperCall, Switch, TaggedTemplate, TemplateLit, Throw, Try, UnaryOp, WhileLoop, Yield,
+};
+use crate::syntax::ast::{op, Const, Node};
+use boa_interner::Sym;
+
+/// `Visitor`s "walk" the AST at certain nodes. Useful for when you need to modify source code, but at
+/// the API level.
+///
+/// `Visitor`s have default implementations which are "reasonable", which is to say that users merely
+/// need to implement the visitor for the specific type they wish to visit. For example, let's say
+/// we want to replace all variables named "y" with a variable named "z".
+///
+/// To do so, we implement `Visitor` over a `NameReplacer` struct, then use `NameReplacer` to, well,
+/// replace the names in a parsed source.
+///
+/// ```
+/// use boa_engine::syntax::ast::node::visit::Visitor;
+/// use boa_engine::Context;
+/// use boa_interner::Sym;
+///
+/// struct NameReplacer {
+///     find: Sym,
+///     replace: Sym,
+/// }
+/// impl<'ast> Visitor<'ast> for NameReplacer {
+///     type Output = ();
+///     type Error = ();
+///
+///     fn visit_sym_mut(&mut self, n: &'ast mut Sym) -> Result<Self::Output, Self::Error> {
+///         if *n == self.find {
+///             *n = self.replace;
+///         }
+///         Ok(())
+///     }
+///
+///     fn get_default_ok() -> Result<Self::Output, Self::Error> {
+///         Ok(())
+///     }
+/// }
+///
+/// // -- snip -- //
+/// # fn main() -> Result<(), ()>{
+/// let source = r#"
+/// let x = 5;
+/// let y = 6;
+/// x + y;
+/// "#;
+/// let mut ctx = Context::default();
+/// let mut parsed = ctx.parse(source).unwrap();
+/// let find = ctx.interner_mut().get_or_intern_static("y");
+/// let replace = ctx.interner_mut().get_or_intern_static("z");
+/// let mut replacer = NameReplacer { find, replace };
+/// replacer.visit_statement_list_mut(&mut parsed)?;
+/// let replaced = parsed.to_indented_string(ctx.interner(), 0);
+/// assert_eq!(
+///     replaced,
+///     r#"
+/// let x = 5;
+/// let z = 6;
+/// x + z;
+/// "#
+///     .trim_start()
+/// );
+/// # Ok(())
+/// # }
+/// ```
+pub trait Visitor<'ast> {
+    type Output;
+    type Error;
+
+    fn visit_node(&mut self, n: &'ast Node) -> Result<Self::Output, Self::Error> {
```

Comment:
These functions would need some documentation, so that they show up in the implementing types, and users know how to implement the trait, especially being `pub`. All methods would need an explanation, in case they want to be used by external crates.